### PR TITLE
Add server extension

### PR DESCRIPTION
**Pull request by carlesarnal**
Many thanks for submitting your Pull Request :heart:!

<details>
<summary>
Generating server stubs from OpenAPI definitions.
</summary>
This PR adds support for generating server stubs from the OpenAPI definition.
</details>

### CODE REVIEW

**Comment by carlesarnal** [comment]
@ricardozanini this is the extension wrapping the Apicurio codegen feature for generating server stubs. I'm opening this in draft so we can discuss what's the best structure since I believe the submitted one is not the best possible (maybe having two folders `server/client` like the `server` one I created?). Let me know what you think!

**Comment by ricardozanini** [comment]
@carlesarnal I think we can have a `client` and `server` top-level modules. Will work better. My only concern is about the maven artifact ID. We should append `-server` to the new one and keep the client as it is to avoid compatibility issues. As we go, we can deprecate one in favor of a `-client` one.

**Comment by carlesarnal** [comment]
> @carlesarnal I think we can have a `client` and `server` top-level modules. Will work better. My only concern is about the maven artifact ID. We should append `-server` to the new one and keep the client as it is to avoid compatibility issues. As we go, we can deprecate one in favor of a `-client` one.

Sorry for the late reply, I've been off for a while. Ok, do you want me to go ahead and do the refactor (essentially moving everything into a `client` module and adding the trailing -server) or you prefer to do it?

Thanks!

**Comment by carlesarnal** [comment]
> > @carlesarnal I think we can have a `client` and `server` top-level modules. Will work better. My only concern is about the maven artifact ID. We should append `-server` to the new one and keep the client as it is to avoid compatibility issues. As we go, we can deprecate one in favor of a `-client` one.
> 
> Sorry for the late reply, I've been off for a while. Ok, do you want me to go ahead and do the refactor (essentially moving everything into a `client` module and adding the trailing -server) or you prefer to do it?
> 
> Thanks!

@ricardozanini I have rebased main on this PR and created a separate `client` directory for the existing extension. Let me know if this is what you had in mind. Thanks!

File: pom.xml
Comment (ricardozanini) [inline pom.xml:18]:
```suggestion
    <module>client</module>
    <module>server</module>
```

Then we create a pom in `client/pom.xml`. 👍

**Review by ricardozanini — COMMENTED** [review]
Thanks @carlesarnal! Yeah, I believe we are on the right path.

**Comment by carlesarnal** [comment]
@ricardozanini CI is green and I have added the client pom. I think this is ready to go, let me know what you think. Thanks!

File: client/docs/pom.xml
Comment (ricardozanini) [inline client/docs/pom.xml:36]: Where is the documentation for the server part? We might need to review the README to add links to both. The client side has been migrated to quarkiverse: https://docs.quarkiverse.io/quarkus-openapi-generator/dev/index.html

We might either:

1. Review this doc to include "Quarkus OpenAPI Generator - Client"
2. Have a unique docs dir with both docs

If you already have docs, we can link both (client and server) in the readme, but still 1 should be done here. wdyt?

**Review by ricardozanini — COMMENTED** [review]
Looks great! I'm just missing the docs, otherwise I think we are good to go.

File: client/docs/pom.xml
Comment (carlesarnal) [reply client/docs/pom.xml:36]: We don't have docs yet (I'm working on having something decent, but things keep happening :(). I have updated the reference in the doc you mentioned. If you're happy with the structure proposed, maybe it's better to merge this PR and then add the docs in a follow-up one? wdyt?

**Review by hbelmiro — COMMENTED** [review]
I wonder if the `client` directory will get the changes from the main branch. If that's not the case, we will lose the most recent changes when we merge this PR.

Also, we need to backport this to the `quarkus2` branch, otherwise, it will become a nightmare backporting other changes to `quarkus2`. I believe this backport will not be as smooth as usual, so I think rather than backporting it after the merge, as we usually do, we should have two simultaneous PRs and merge both only when all tests are passing.

**Comment by ricardozanini** [comment]
@hbelmiro or we can wait to merge this one until we drop quarkus 2, which will be soon, likely in 3 weeks maybe. wdyt? @carlesarnal cc

**Comment by carlesarnal** [comment]
> @hbelmiro or we can wait to merge this one until we drop quarkus 2, which will be soon, likely in 3 weeks maybe. wdyt? @carlesarnal cc

I think that probably makes more sense than having to backport this to quarkus2.

File: client/deployment/pom.xml
Comment (fjtirado) [inline client/deployment/pom.xml:6]: I think we should keep the artifact name untouched for backward compatibilitty reasons and just add the server suffix to the new artifact containing the server stubs

**Review by fjtirado — CHANGES_REQUESTED** [review]
Lets keep the existing client pom artifacts ids existing names to avoid backward compatibiity issues. 
Everything else looks superb. Fantastic work, congrats.

**Review by fjtirado — APPROVED** [review]

**Comment by carlesarnal** [comment]
I have updated this PR from the changes from main so being outdated does not prevent it from being merged. If anything else is needed just let me know.

**Comment by fjtirado** [comment]
@carlesarnal I merged the PR from @mcruzdev, try resolving the conflict and hopefully the PR will be green now

**Comment by carlesarnal** [comment]
> @carlesarnal I merged the PR from @mcruzdev, try resolving the conflict and hopefully the PR will be green now

PR updated and conflicts resolved.

**Comment by mcruzdev** [comment]
@carlesarnal the branch has conflicts, can you solve?  I do not know why but the `
Build Website / Build Website (pull_request) ` does not work well, if it persist after solving the conflicts I will take a look.

**Comment by carlesarnal** [comment]
> @carlesarnal the branch has conflicts, can you solve? I do not know why but the `Build Website / Build Website (pull_request)` does not work well, if it persist after solving the conflicts I will take a look.

I have solved the conflicts but it's still failing.

File: server/deployment/src/main/java/io/quarkiverse/openapi/server/generator/deployment/codegen/ApicurioCodegenWrapper.java
Comment (hbelmiro) [inline server/deployment/src/main/java/io/quarkiverse/openapi/server/generator/deployment/codegen/ApicurioCodegenWrapper.java:27]: You're getting the wrong logger.

```suggestion
    private static final Logger log = LoggerFactory.getLogger(ApicurioCodegenWrapper.class);
```

File: server/deployment/src/main/java/io/quarkiverse/openapi/server/generator/deployment/codegen/ApicurioCodegenWrapper.java
Comment (hbelmiro) [inline server/deployment/src/main/java/io/quarkiverse/openapi/server/generator/deployment/codegen/ApicurioCodegenWrapper.java:30]: This attribute is never used and can be removed.

File: server/deployment/src/main/java/io/quarkiverse/openapi/server/generator/deployment/codegen/ApicurioCodegenWrapper.java
Comment (hbelmiro) [inline server/deployment/src/main/java/io/quarkiverse/openapi/server/generator/deployment/codegen/ApicurioCodegenWrapper.java:48]:
```suggestion
        log.info("Generating JAX-RS interfaces and beans from: {}", openApiResource);
```

File: server/deployment/src/main/java/io/quarkiverse/openapi/server/generator/deployment/codegen/ApicurioCodegenWrapper.java
Comment (hbelmiro) [inline server/deployment/src/main/java/io/quarkiverse/openapi/server/generator/deployment/codegen/ApicurioCodegenWrapper.java:75]:
```suggestion
        }
```

File: server/deployment/src/main/java/io/quarkiverse/openapi/server/generator/deployment/codegen/ApicurioOpenApiServerCodegen.java
Comment (hbelmiro) [inline server/deployment/src/main/java/io/quarkiverse/openapi/server/generator/deployment/codegen/ApicurioOpenApiServerCodegen.java:57]: Shouldn't we throw the exception and abort the generation?

**Review by hbelmiro — COMMENTED** [review]
Thank you @carlesarnal. Amazing contribution.
I left a few comments and have one question: 

Does Apicurio generate code for RESTEasy Reactive, Classic, or both (user chooses)? On the client side, we generate one or another, based on the dependencies the users added to their poms. We do that to keep backward compatibility since we had only Classic support in the beginning. But supporting both implementations brings us more complexity. In my opinion, we could support only the Reactive implementation on the server side.

File: server/deployment/src/main/java/io/quarkiverse/openapi/server/generator/deployment/codegen/ApicurioOpenApiServerCodegen.java
Comment (carlesarnal) [reply server/deployment/src/main/java/io/quarkiverse/openapi/server/generator/deployment/codegen/ApicurioOpenApiServerCodegen.java:57]: Not really. This is processing all the spec files in the directory, what should be done for sure is to log this properly.

**Comment by carlesarnal** [comment]
> Thank you @carlesarnal. Amazing contribution. I left a few comments and have one question:
> 
> Does Apicurio generate code for RESTEasy Reactive, Classic, or both (user chooses)? On the client side, we generate one or another, based on the dependencies the users added to their poms. We do that to keep backward compatibility since we had only Classic support in the beginning. But supporting both implementations brings us more complexity. In my opinion, we could support only the Reactive implementation on the server side.

Thanks for the feedback! 

So, this PR only includes support for classic, as you have probably already realized, but Apicurio Codegen has the capability for generating reactive as well. As for which one to pick if you want one or another in the quarkus extension, that is up to you since for us, in this case, it's relatively easy to do the switch (it's a parameter in the configuration that can be just a configuration property in the `application.properties`, obviously autodetecting it is a bit more tricky).

Let me know what you think. Maybe we can have this support and I can submit a followup PR with reactive support.

**Comment by hbelmiro** [comment]
Wouldn't it be better to start with reactive support then? We can add classic support later only if needed since classic is kind of deprecated.
If we start with classic, it's almost certain that we'll have to add reactive support soon.
At the same time, I don't want to block the merge.
So, what do you think @carlesarnal and @ricardozanini? Would switching to reactive in this PR require much work?

**Comment by akram** [comment]
Hi folks and @carlesarnal , great cocntribution. Really interested in seeing it merged.

**Comment by ricardozanini** [comment]
@carlesarnal can you please rebase and open a follow up issue to introduce reactive? @hbelmiro I'd like to merge this sooner rather than later. We can handle the other use cases in future issues.

File: server/docs/modules/ROOT/pages/index.adoc
Comment (mcruzdev) [inline server/docs/modules/ROOT/pages/index.adoc:5]:
```suggestion
WARNING: This is the instructions for the latest SNAPSHOT version (main branch). Please, see the [latest **released** documentation](https://docs.quarkiverse.io/quarkus-openapi-generator/dev/index.html) if you are looking for instructions.
```

**Comment by carlesarnal** [comment]
Sorry, I was on PTO. @ricardozanini rebase done and issue created. Let me know if anything else is needed.

**Review by hbelmiro — APPROVED** [review]

**Review by mcruzdev — APPROVED** [review]

**Comment by ricardozanini** [comment]
@allcontributors please add @carlesarnal for maintenance

**Comment by markuszoeller** [comment]
@hbelmiro @mcruzdev I'd like to give this a try as a user. What's the next planned release? Am I right in the assumption that I would see an entry called `quarkus-openapi-generator-server` at https://mvnrepository.com/artifact/io.quarkiverse.openapi.generator when this is released?

I'm building with `gradle` and tried this a few minutes ago:
```build.gradle
// file: build.gradle

dependencies {
  // [...]
  implementation "io.quarkiverse.openapi.generator:quarkus-openapi-generator-server:3.0.0-SNAPSHOT"
}
```

But the build failed with this:
```shell
$ ./gradlew clean build

FAILURE: Build failed with an exception.

* What went wrong:
Could not determine the dependencies of task ':compileJava'.
> Could not resolve all dependencies for configuration ':quarkusDependency'.
   > Could not find io.quarkiverse.openapi.generator:quarkus-openapi-generator-server:3.0.0-SNAPSHOT.
     Searched in the following locations:
       - https://repo.maven.apache.org/maven2/io/quarkiverse/openapi/generator/quarkus-openapi-generator-server/3.0.0-SNAPSHOT/maven-metadata.xml
       - https://repo.maven.apache.org/maven2/io/quarkiverse/openapi/generator/quarkus-openapi-generator-server/3.0.0-SNAPSHOT/quarkus-openapi-generator-server-3.0.0-SNAPSHOT.pom
       - file:/Users/markus/.m2/repository/io/quarkiverse/openapi/generator/quarkus-openapi-generator-server/3.0.0-SNAPSHOT/maven-metadata.xml
       - file:/Users/markus/.m2/repository/io/quarkiverse/openapi/generator/quarkus-openapi-generator-server/3.0.0-SNAPSHOT/quarkus-openapi-generator-server-3.0.0-SNAPSHOT.pom
```

**Comment by ricardozanini** [comment]
@carlesarnal can you give us a hand?

**Comment by ricardozanini** [comment]
@markuszoeller many thanks for being the guinea pig. :)

So if we have a trusted stable codebase for the server plugin, I can release a major version.

**Comment by carlesarnal** [comment]
> @carlesarnal can you give us a hand?

I'll look at this ASAP.

**Comment by carlesarnal** [comment]
> > @carlesarnal can you give us a hand?
> 
> I'll look at this ASAP.

@markuszoeller  Until there is a new release making the artifact available in maven central as you described, you need the artifact to be present in your local repository. If you add the dependency straight to a `build.gradle` it will fail but then if you build the project locally (just a `mvn clean install`) it adds the artifact to your local repository. After this step, it builds the project with no issues (just tried it with an example gradle project).

**Comment by markuszoeller** [comment]
> Until there is a new release making the artifact available in maven central as you described, you need the artifact to be present in your local repository. If you add the dependency straight to a `build.gradle` it will fail but then if you build the project locally (just a `mvn clean install`) it adds the artifact to your local repository. After this step, it builds the project with no issues (just tried it with an example gradle project).

This worked for me, thank you! The generated code looks very good so far. We probably test this over the next weeks. If I notice something odd then I'm going to open a GH Issue.

**Comment by ygyg70** [comment]
Checking this feature, I have an OAS file that can generate a client, and also generate server using other libraries, however it fails with the exception below. Any tip how to debug is appreciated.

```
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal io.quarkus.platform:quarkus-maven-plugin:3.7.4:generate-code (default) on project mock-solve: Quarkus code generation phase has failed
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2 (MojoExecutor.java:347)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:330)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:213)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:175)
    at org.apache.maven.lifecycle.internal.MojoExecutor.access$000 (MojoExecutor.java:76)
    at org.apache.maven.lifecycle.internal.MojoExecutor$1.run (MojoExecutor.java:163)
    at org.apache.maven.plugin.DefaultMojosExecutionStrategy.execute (DefaultMojosExecutionStrategy.java:39)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:160)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:105)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:73)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:53)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:118)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:260)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:172)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:100)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:821)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:270)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
    at jdk.internal.reflect.DirectMethodHandleAccessor.invoke (DirectMethodHandleAccessor.java:103)
    at java.lang.reflect.Method.invoke (Method.java:580)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: org.apache.maven.plugin.MojoExecutionException: Quarkus code generation phase has failed
    at io.quarkus.maven.GenerateCodeMojo.generateCode (GenerateCodeMojo.java:94)
    at io.quarkus.maven.GenerateCodeMojo.doExecute (GenerateCodeMojo.java:54)
    at io.quarkus.maven.QuarkusBootstrapMojo.execute (QuarkusBootstrapMojo.java:171)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:126)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2 (MojoExecutor.java:342)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:330)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:213)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:175)
    at org.apache.maven.lifecycle.internal.MojoExecutor.access$000 (MojoExecutor.java:76)
    at org.apache.maven.lifecycle.internal.MojoExecutor$1.run (MojoExecutor.java:163)
    at org.apache.maven.plugin.DefaultMojosExecutionStrategy.execute (DefaultMojosExecutionStrategy.java:39)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:160)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:105)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:73)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:53)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:118)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:260)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:172)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:100)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:821)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:270)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
    at jdk.internal.reflect.DirectMethodHandleAccessor.invoke (DirectMethodHandleAccessor.java:103)
    at java.lang.reflect.Method.invoke (Method.java:580)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: java.lang.reflect.InvocationTargetException
    at jdk.internal.reflect.DirectMethodHandleAccessor.invoke (DirectMethodHandleAccessor.java:118)
    at java.lang.reflect.Method.invoke (Method.java:580)
    at io.quarkus.maven.GenerateCodeMojo.generateCode (GenerateCodeMojo.java:88)
    at io.quarkus.maven.GenerateCodeMojo.doExecute (GenerateCodeMojo.java:54)
    at io.quarkus.maven.QuarkusBootstrapMojo.execute (QuarkusBootstrapMojo.java:171)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:126)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2 (MojoExecutor.java:342)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:330)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:213)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:175)
    at org.apache.maven.lifecycle.internal.MojoExecutor.access$000 (MojoExecutor.java:76)
    at org.apache.maven.lifecycle.internal.MojoExecutor$1.run (MojoExecutor.java:163)
    at org.apache.maven.plugin.DefaultMojosExecutionStrategy.execute (DefaultMojosExecutionStrategy.java:39)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:160)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:105)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:73)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:53)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:118)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:260)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:172)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:100)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:821)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:270)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
    at jdk.internal.reflect.DirectMethodHandleAccessor.invoke (DirectMethodHandleAccessor.java:103)
    at java.lang.reflect.Method.invoke (Method.java:580)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: java.lang.StackOverflowError
    at com.fasterxml.jackson.databind.node.ObjectNode.set (ObjectNode.java:587)
    at io.apicurio.datamodels.models.util.JsonUtil.setStringProperty (JsonUtil.java:577)
    at io.apicurio.datamodels.models.openapi.v31.io.OpenApi31ModelWriter.writeSchema (OpenApi31ModelWriter.java:1079)
    at io.apicurio.datamodels.models.openapi.v31.io.OpenApi31ModelWriter.writeSchema (OpenApi31ModelWriter.java:1112)
    at io.apicurio.datamodels.models.openapi.v31.io.OpenApi31ModelWriter.lambda$writeSchema$58 (OpenApi31ModelWriter.java:1134)
    at java.util.LinkedHashMap$LinkedKeySet.forEach (LinkedHashMap.java:729)
    at io.apicurio.datamodels.models.openapi.v31.io.OpenApi31ModelWriter.writeSchema (OpenApi31ModelWriter.java:1132)
    at io.apicurio.datamodels.models.openapi.v31.io.OpenApi31ModelWriter.writeSchema (OpenApi31ModelWriter.java:1112)
    at io.apicurio.datamodels.models.openapi.v31.io.OpenApi31ModelWriter.lambda$writeSchema$58 (OpenApi31ModelWriter.java:1134)
    at java.util.LinkedHashMap$LinkedKeySet.forEach (LinkedHashMap.java:729)
...
```

**Comment by ygyg70** [comment]
The generated code included lines such as the one below, where "Schema-" is an extra, and therefore doesn't compile:
```private List<Schemas-Tag> tag = new ArrayList<Schemas-Tag>();```
Any ideas?

**Comment by carlesarnal** [comment]
Hello @ygyg70 if you have a problem, please, open a new issue (this is the PR for introducing the support). Also, be sure to include some information to reproduce the problem.

**Pull request merged**
